### PR TITLE
Bridge handling improved (bsc#962824)

### PR DESCRIPTION
**Pull request by teclator**
I have unified the configuration overview with bridges involved with the bonding one as you can see in the screenshots. There have been fixed also some small issue.

## Bond overview
enslavedbond

## Bridge overview
enslavedbridge

## Old bridge configuration
old_bridge

## New bridge configuration
new_bridge

### CODE REVIEW

File: src/modules/Lan.rb [hunk 1]
```diff
@@ -1039,7 +1042,6 @@ def Packages
         # for wlan require iw instead of wireless-tools (bnc#539669)
         "wlan" => "iw",
         "vlan" => "vlan",
-        "br"   => "bridge-utils",
```
Comment (teclator) [inline src/modules/Lan.rb]: As we are using iptools2 and wicked, this dependency should be removed as this way of configure bridges is currently deprecated.

File: src/include/network/lan/address.rb [hunk 2]
```diff
@@ -1553,8 +1555,6 @@ def AddressDialog
         LanItems.vlan_id = Builtins.tostring(
           Ops.get_integer(@settings, "VLAN_ID", 0)
         )
-      elsif LanItems.type == "br"
-        LanItems.bridge_ports = Ops.get_string(@settings, "BRIDGE_PORTS", "")
```
Comment (teclator) [inline src/include/network/lan/address.rb]: Removing this to `StoreBridge` avoid current ports in case that the `bridge ports tab` is not visited.
## Previously

bridge_ports_removed

## After the fix

bridge_ports_removed

File: src/include/network/lan/address.rb [hunk 3]
```diff
@@ -1355,6 +1355,19 @@ def AddressDialog
...
 
+      if LanItems.type == "br"
+        ports = LanItems.bridge_ports.split(" ").select { |p| old_bridge_config?(p) }
+
+        unless ports.empty?
+          message = format(_("The bridge ports listed below are configured in the old way.\n\n" \
+                            "Bridge ports: %s\n\n" \
+                            "Do you want to adapt them now?"), ports.join(", "))
+          if Popup.YesNoHeadline(Label.WarningMsg, message)
+            ports.each { |p| configure_as_bridge_port(p) }
```
Comment (teclator) [inline src/include/network/lan/address.rb]: We are just adapting it when the bridge is edited as we can see in this gif:

ask_for_adapt


We could also ask for adapt the configuration when the enslaved interface is edited.

File: package/yast2-network.changes [hunk 4]
```diff
@@ -1,4 +1,13 @@
 -------------------------------------------------------------------
+Wed Sep 21 11:56:38 UTC 2016 - kanderssen@suse.com
+
+- Bridge handling has been improved (bsc#962824).
+  - "NONE" is shown instead of 0.0.0.0 for old bridge configuration
+  - The bridge master is shown in the enslaved interface.
+  - The interfaces overview is updated after a bridge is modified
+- 3.1.170.1
```
Comment (mchf) [inline package/yast2-network.changes]: if I remember correctly, SP3 didn't diverged so far.

Comment (teclator) [inline package/yast2-network.changes]: This change https://github.com/yast/yast-network/pull/440 was the reason I choose that version.

File: src/include/network/complex.rb [hunk 5]
```diff
@@ -401,18 +401,17 @@ def DeviceStatus(devtype, devname, devmap)
...
         # Abbreviation for "The interface is Managed by NetworkManager"
         return _("Managed")
       end
-      ip = Ops.get_string(devmap, "BOOTPROTO", "static")
-      ip = if ip.nil? || ip == "" || ip == "static"
-        Ops.get_string(devmap, "IPADDR", "")
+      bootproto = devmap["BOOTPROTO"] || "static"
+
+      if bootproto.empty? || bootproto == "static"
+        devmap["IPADDR"]
```
Comment (teclator) [inline src/include/network/complex.rb]: In case of `nil` `bootproto` is set to `"static"` as you can see in line 408. Now we have a unit test for `DeviceProtocol` method and we can see that for `nil`, `""` and `"static"` we return the IP address.

File: src/include/network/lan/address.rb [hunk 6]
```diff
@@ -1596,6 +1593,10 @@ def AddressDialog
 
   private
 
+    def get_selected_bridges
```
Comment (mchf) [inline src/include/network/lan/address.rb]: missing documentation for new method

File: src/include/network/lan/address.rb [hunk 7]
```diff
@@ -475,25 +475,11 @@ def InitBridge(key)
...
-            " "
-          )
-        )
-      )
-      Builtins.y2milestone(
-        "store bridge %1 : %2",
-        key,
-        Ops.get_string(@settings, "BRIDGE_PORTS", "")
-      )
+      @settings["BRIDGE_PORTS"] = get_selected_bridges.join(" ")
```
Comment (mchf) [inline src/include/network/lan/address.rb]: misleading function name. Something like ```selected_bridge_ports``` would be better in my POV.

File: src/include/network/lan/address.rb [hunk 8]
```diff
@@ -1369,6 +1355,19 @@ def AddressDialog
         Ops.set(@settings, "VLAN_ID", Builtins.tointeger(LanItems.vlan_id))
       end
 
+      if LanItems.type == "br"
```
Comment (mchf) [inline src/include/network/lan/address.rb]: i think that this branch deserves own method

File: src/include/network/lan/bridge.rb [hunk 9]
```diff
@@ -34,16 +34,43 @@ def initialize_network_lan_bridge(_include_target)
...
+      ip == "0.0.0.0" ? "" : ip.to_s
+    end
+
+    # Checks if the given interface is using the old way of config enslaved
+    # interfaces with bootproto as static and 0.0.0.0 as IPADDR
+    #
+    # @param [String] interface name
+    # @return [Boolean] returns true if given enslaved interface is configured
+    # in the old way
+    def old_bridge_config?(ifcfg_name)
```
Comment (mchf) [inline src/include/network/lan/bridge.rb]: no bridge but bridge port config

File: src/include/network/lan/address.rb (hunk 2 above)
Comment (mchf) [inline src/include/network/lan/address.rb]: nice ;-) Very common pattern in network code :-/

File: src/modules/LanItems.rb [hunk 10]
```diff
@@ -1311,25 +1339,26 @@ def BuildLanOverview
...
-            ifcfg_name,
-            ifcfg_conf
-          )
+          status = DeviceStatus(ifcfg_type, ifcfg_name, ifcfg_conf)
 
           bullets << _("Device Name: %s") % ifcfg_name
           bullets += startmode_overview(key)
+
+          if bridge_index[ifcfg_name]
+            ip = bridge_ip(ip).empty? ? "NONE" : bridge_ip(ip)
```
Comment (mchf) [inline src/modules/LanItems.rb]: AFAIK this is not correct - bridge ports must not have ip addr (I mean other than 0.0.0.0 in case of pre-wicked configs)

File: src/modules/LanItems.rb [hunk 11]
```diff
@@ -2382,6 +2406,34 @@ def self.publish_variable(name, type)
...
+
+      format("%s: %s", desc, slaves.join(" "))
+    end
+
+    # Check if the given interface is enslaved in a bond or in a bridge
+    #
+    # @return [Boolean] true if enslaved
+    def enslaved?(ifcfg_name)
+      bond_index = BuildBondIndex()
+      return true if bond_index[ifcfg_name] || bridge_index[ifcfg_name]
```
Comment (mchf) [inline src/modules/LanItems.rb]: I'd keep these separated.

File: src/include/network/lan/address.rb (hunk 8 above)
Comment (teclator) [inline src/include/network/lan/address.rb]: I was thinking also on it so just need to decide how to name it

File: src/modules/LanItems.rb (hunk 10 above)
Comment (teclator) [inline src/modules/LanItems.rb]: If the interface is edited before removed from the bridge for example, in that case the IP will be shown. So in this case we are just replacing `0.0.0.0` in other case there could be some human error or not expected configuration and for that should not be hidden anyway.

File: src/modules/LanItems.rb (hunk 11 above)
Comment (teclator) [inline src/modules/LanItems.rb]: You always can check directly if it's enslaved in a bridge or in a bond checking the associated hash, so in some way it is separated and this method was just to check if we have to add something to the Note column reusing code, but I can separate it if it makes the code more clear and clean.

Comment (mchf) [inline src/modules/LanItems.rb]: I know. I just thought that it is a bit overkill to ask for the reverse mapping for bond when in fact checking for bridge and vice versa. One of those is always useless the device can be either bridged or bonded, never both.

File: src/modules/LanItems.rb (hunk 10 above)
Comment (mchf) [inline src/modules/LanItems.rb]: if the interface is edited before removing from bridge, then yast should remove it from bridge. Otherwise you could put system into inconsistent / incorrect state.

File: src/include/network/lan/address.rb (hunk 8 above)
Comment (mchf) [inline src/include/network/lan/address.rb]: yeah .. the hardest part ;-)

File: src/include/network/complex.rb (hunk 5 above)
Comment (mchf) [inline src/include/network/complex.rb]: I've addressed ```devmap["IPADDR"]``` here

File: package/yast2-network.changes (hunk 4 above)
Comment (mchf) [inline package/yast2-network.changes]: ok, i didn't remember well ;-)

File: src/modules/LanItems.rb (hunk 10 above)
Comment (teclator) [inline src/modules/LanItems.rb]: @mchf of course this is also true, but i mean, if there is also ans IP by whatever reason, for example manual edition of the file, then there could be a wrong IP, what to do in that case, show the IP, show ERROR.

The values should be `0.0.0.0` or `NONE` in all the cases, but we never know, and in that unknown scenarios, how you prefer to handle it?

File: src/modules/LanItems.rb (hunk 11 above)
Comment (teclator) [inline src/modules/LanItems.rb]: So, as you prefer I can remove that method and rewrite the `Note` part of code checking both indexes and writing the note column according to them.

File: src/include/network/lan/address.rb [hunk 12]
```diff
@@ -1593,14 +1546,72 @@ def AddressDialog
 
   private
 
-    def get_selected_bridges
+    # Checks if the current LanItem is a bridge or an enslaved interface part
+    # of a bridge. In case that some configuration is not correct then adapt it
+    # if the user accepts.
+    def check_bridge_ports_config
+      return false unless LanItems.type == "br"
```
Comment (mchf) [inline src/include/network/lan/address.rb]: As I try to drop dependency on these "global" status variables I prefer to have new functions stateless (e.g. checking device type on-the-fly as you never know who forgot to reset or set it somewhere to something reasonable). It should also ease "big bang" refactoring of ```LanItems``` one day.

File: src/include/network/lan/address.rb [hunk 13]
```diff
@@ -1593,14 +1546,72 @@ def AddressDialog
...
+
+      ports = LanItems.bridge_ports.split(" ").select { |p| old_bridge_port_config?(p) }
+
+      unless ports.empty?
+        ports.each { |p| configure_as_bridge_port(p) } if fix_bridge_port_config?(ports)
```
Comment (mchf) [inline src/include/network/lan/address.rb]: you really want to raise a popup for every old fashioned bridge port? Poor SLE-11 user with 10 ports bridge ;-)

File: src/include/network/lan/bridge.rb [hunk 14]
```diff
@@ -60,6 +87,21 @@ def configure_as_bridge_port(device)
 
       NetworkInterfaces.Commit
       NetworkInterfaces.Add
+
+      NetworkInterfaces.Current = selected_interface
+    end
+
+    # Asks the user about adapt the current bridge port config in case that it
+    # is configured in the old way (BOOTPROTO == static) & (IPADDR == 0.0.0.0).
+    def fix_bridge_port_config?(ports)
```
Comment (teclator) [inline src/include/network/lan/bridge.rb]: Maybe `adapt_bridge_port_config?` could be a better name for this method, and then instead of configure_as_bridge_port define this method as it is an adaption and not a new configuration at all:

```
def adapt_bridge_port_config?(ports)
  ...
end
 
def adapt_bridge_port_config!(ifcfg_name)
  item_id = LanItems.get_configured(ifcfg_name)
  devmap = LanItems.GetDeviceMap(item_id)

  return false unless devmap

  devmap["IPADDR"] = ""
  devmap["NETMASK"] = ""
  devmap["BOOTPROTO"] = "none"
  # take out PREFIXLEN from old configuration (BNC#735109)
  devmap["PREFIXLEN"] = ""

  LanItems.SetDeviceMap(item_id, devmap)

  true
end
```

@mchf what do you think?

File: src/include/network/complex.rb (hunk 5 above)
Comment (teclator) [inline src/include/network/complex.rb]: true .. :)

File: src/include/network/lan/address.rb (hunk 13 above)
Comment (teclator) [inline src/include/network/lan/address.rb]: It is only raised once with all the ports in the question

File: src/include/network/complex.rb (hunk 5 above)
Comment (teclator) [inline src/include/network/complex.rb]: fixed

File: src/include/network/lan/address.rb (hunk 6 above)
Comment (teclator) [inline src/include/network/lan/address.rb]: Added documentation

File: src/include/network/lan/address.rb (hunk 7 above)
Comment (teclator) [inline src/include/network/lan/address.rb]: Renamed and also renamed get_selected_slaves

File: src/include/network/lan/bridge.rb (hunk 9 above)
Comment (teclator) [inline src/include/network/lan/bridge.rb]: Renamed

File: src/include/network/lan/address.rb (hunk 12 above)
Comment (teclator) [inline src/include/network/lan/address.rb]: This was exactly the reason I proposed the new method for adapt the config, because the other ones depends on the NetworkInterfaces instance, and it also changed the Current to the last port edited what could produce some unexpected issues.

File: src/include/network/complex.rb (hunk 5 above)
Comment (mchf) [inline src/include/network/complex.rb]: thx

File: src/modules/LanItems.rb (hunk 10 above)
Comment (mchf) [inline src/modules/LanItems.rb]: well if there is no (```IPADDR=0.0.0.0``` and ```BOOTPROTO=static```) or (```BOOTPROTO=none``` and possibly ```IPADDR=<whatever>```) then wicked won't accept that anyway. So in first case is having an ip set invalid and we shouldn't mask it, in second case it doesn't depend and we do not need to take care of it.

File: src/modules/LanItems.rb (hunk 11 above)
Comment (mchf) [inline src/modules/LanItems.rb]: well, my idea behind this is that one day there will be a generic ```iface_type``` class with ```bridge_type``` and ```bond_type``` successors. If you rewrite it into two, we can copy paste methods as they are into these classes with almost no change.

If it makes sense to you split it. If you think that I'm overengineering things keep it as it is.

Comment (teclator) [inline src/modules/LanItems.rb]: In the future could be as simple as ```"Enslaved in #{interface.master.name}" if interface.master`` or something like that

File: src/include/network/lan/address.rb (hunk 12 above)
Comment (mchf) [inline src/include/network/lan/address.rb]: Just thinking ...
... using ```NetworkInterfaces.GetType``` makes no sense when creating new bridge until new configuration is commited into ```NetworkInterfaces```
... so you can just move the type check out of this method.

File: src/include/network/lan/address.rb (hunk 13 above)
Comment (mchf) [inline src/include/network/lan/address.rb]: sure, my fault.

File: src/include/network/lan/bridge.rb [hunk 15]
```diff
@@ -34,16 +34,43 @@ def initialize_network_lan_bridge(_include_target)
...
+      selected_interface = NetworkInterfaces.Current
       log.info("Adapt device #{device} as bridge port")
 
       # when using wicked every device which can be bridged
       # can be set to BOOTPROTO=none. No workaround with
       # BOOTPROTO=static required anymore
-      NetworkInterfaces.Edit(device)
+      unless NetworkInterfaces.Edit(device)
+        Lan.autoconf_slaves += [device] unless Lan.autoconf_slaves.include? device
+        return false
```
Comment (mchf) [inline src/include/network/lan/bridge.rb]: inconsistent return value

File: src/include/network/lan/bridge.rb [hunk 16]
```diff
@@ -60,6 +87,21 @@ def configure_as_bridge_port(device)
 
       NetworkInterfaces.Commit
       NetworkInterfaces.Add
+
+      NetworkInterfaces.Current = selected_interface
```
Comment (mchf) [inline src/include/network/lan/bridge.rb]: why to return original ```Current```?

File: src/include/network/lan/bridge.rb (hunk 14 above)
Comment (mchf) [inline src/include/network/lan/bridge.rb]: well it better describes current use case, but i don't see any benefit in having such one purpose method. So do you have any other pros? ;-)

File: src/include/network/lan/bridge.rb [hunk 17]
```diff
@@ -34,16 +34,43 @@ def initialize_network_lan_bridge(_include_target)
...
     def configure_as_bridge_port(device)
+      selected_interface = NetworkInterfaces.Current
       log.info("Adapt device #{device} as bridge port")
 
       # when using wicked every device which can be bridged
       # can be set to BOOTPROTO=none. No workaround with
       # BOOTPROTO=static required anymore
-      NetworkInterfaces.Edit(device)
+      unless NetworkInterfaces.Edit(device)
+        Lan.autoconf_slaves += [device] unless Lan.autoconf_slaves.include? device
```
Comment (mchf) [inline src/include/network/lan/bridge.rb]: well this autoconf stuff is part of a fate request for bonding devices where it was requested to do an udev update for slaves. As it was not requested for bridges, it shouldn't be implemented for bridges in my POV.

Comment (teclator) [inline src/include/network/lan/bridge.rb]: Ok, as we was talking some time ago about unifying bonding and bridge management as much as possible was what I tried but this is going too further and some changes should not be necessary and also invalid, so taking this in account I'm going to modify the PR in concordance.

File: test/bridge_test.rb [hunk 18]
```diff
@@ -6,87 +6,167 @@
...
 
-    let(:expected_bridgeable) do
-      [
-        "bond0",
-        "eth4",
-        "eth11",
-        "eth12",
-        "tap0"
-      ]
+  describe "#old_bridge_port_config?" do
```
Comment (mvidner) [inline test/bridge_test.rb]: please add a test for the "eth1" case , which has `  { "BOOTPROTO" => "none" } ` and so should return false

File: test/bridge_test.rb [hunk 19]
```diff
@@ -6,87 +6,167 @@
...
+        "eth5" => { "BOOTPROTO" => "static", "STARTMODE" => "nfsroot" },
+        "eth6" => { "BOOTPROTO" => "static", "STARTMODE" => "ifplugd" }
+      },
+      "tun"  => {
+        "tun0" => {
+          "BOOTPROTO" => "static",
+          "STARTMODE" => "onboot",
+          "TUNNEL"    => "tun"
+        }
+      },
```
Comment (mvidner) [inline test/bridge_test.rb]: This case is not documented in the new yard comment

File: src/include/network/lan/bridge.rb [hunk 20]
```diff
@@ -34,32 +34,97 @@ def initialize_network_lan_bridge(_include_target)
...
+        end
+        NetworkInterfaces.Current["_aliases"] = {}
+
+        NetworkInterfaces.Commit
+        NetworkInterfaces.Add
 
-      # remove all aliases (bnc#590167)
-      aliases = NetworkInterfaces.Current["_aliases"] || {}
-      aliases.each do |alias_name, alias_ip|
-        NetworkInterfaces.DeleteAlias(device, alias_name) if alias_ip
```
Comment (mvidner) [inline src/include/network/lan/bridge.rb]: Sorry, I would be totally baffled as a user, and would not know what to answer. Why are we even asking?

File: src/include/network/lan/address.rb (hunk 3 above)
Comment (mvidner) [inline src/include/network/lan/address.rb]: What is there for the user to decide? IMHO this pop-up is harmful, burdening the user with a decision they cannot understand ("old way, new way"?)
I propose that we simply adapt the config and just log the action.

File: src/include/network/lan/bridge.rb (hunk 15 above)
Comment (mvidner) [inline src/include/network/lan/bridge.rb]: @teclator what is this method supposed to return in this case, and in the normal case? Add a @return tag?

File: src/modules/LanItems.rb (hunk 10 above)
Comment (mvidner) [inline src/modules/LanItems.rb]: The string `"NONE"` is wrong (I guess you have blindly copied it from the bug reporter) because it is not translated and it is not an abbreviation. We should either use `""` ( it will be used in a table cell so that is OK) or `_("none")`.

File: src/include/network/lan/address.rb [hunk 21]
```diff
@@ -475,25 +475,11 @@ def InitBridge(key)
...
-        )
-      )
-      Builtins.y2milestone(
-        "store bridge %1 : %2",
-        key,
-        Ops.get_string(@settings, "BRIDGE_PORTS", "")
-      )
+      @settings["BRIDGE_PORTS"] = selected_bridge_ports.join(" ")
+
+      LanItems.bridge_ports = @settings["BRIDGE_PORTS"]
```
Comment (mvidner) [inline src/include/network/lan/address.rb]: NP: So `@settings["BRIDGE_PORTS"]` is a String but `@settings["SLAVES"]` is an Array. WTF. I know this is a bit out of focus, but could we at least document it and add a FIXME to unify it?

File: src/modules/LanItems.rb [hunk 22]
```diff
@@ -732,7 +732,7 @@ def IsBondable(bondMaster, itemId)
     # @param [Fixnum] itemId        index into LanItems::Items
     # TODO: bridgeMaster is not used yet bcs detection of bridge master
```
Comment (mvidner) [inline src/modules/LanItems.rb]: This TODO is no longer valid

File: src/include/network/lan/address.rb (hunk 3 above)
Comment (teclator) [inline src/include/network/lan/address.rb]: @mvidner 

It was a doubt I have  at the beginning of the PBI as it was also commented in the card. After discussing it with @mchf the conclusion was that automatic conversion does not means not ask the user about it.

We could do it automatically, in that case when? when the interface is edited and saved or also do it although the user press cancel in the edition. If should be converted so why not just have a method that checks current interface configuration when the client is called and adapt old config style?

We just decided to prompt the user just to verify he really wanted to adapt the current configuration and also being a litle conservative taking in account that if something change in current workflow the network is restarted and should not restarted if not (so maybe the user does not expect changes when `OK` pressed but then he realized that the network was restarted and some files changed magically.

File: src/modules/LanItems.rb (hunk 10 above)
Comment (teclator) [inline src/modules/LanItems.rb]: I just used the same was used in bonding that is the BOOTPROTO upcased

https://github.com/yast/yast-network/blob/master/src/modules/LanItems.rb#L1272

So I could adapt both for use `_("none")` in that case, or just with the enslaved note should be enough and just let that column as blank

File: src/include/network/lan/address.rb (hunk 21 above)
Comment (teclator) [inline src/include/network/lan/address.rb]: Sure, and is how differs also in configuration per interface. As bonding master is configured as:
```
BONDING_SLAVE0='eth0'
BONDING_SLAVE1='eth1'
```
while bridge_ports are configured as 
```
BRIDGE_PORTS='eth0 eth1'
```

File: src/include/network/lan/bridge.rb (hunk 15 above)
Comment (teclator) [inline src/include/network/lan/bridge.rb]: At the end i decided to just let it as it was, and it always returns true and just add the changes if `NetworkInterfaces.Edit(device)` success and also remember the previous Interfaces selected by NetworkInterface to avoid let it in a different one what could produce not desired results.

https://github.com/teclator/yast-network/blob/bridge_overview_improvement/src/include/network/lan/bridge.rb#L92

As an alternative I added the adapt_old_port_config method which avoid to work over NetworkInterfaces and where we could also add the logic to write the config to the files without restarting the network

https://github.com/teclator/yast-network/blob/bridge_overview_improvement/src/include/network/lan/bridge.rb#L113

File: src/include/network/lan/bridge.rb (hunk 16 above)
Comment (teclator) [inline src/include/network/lan/bridge.rb]: is not returned anymore, i let it as it was returning true again

**Comment by teclator** [comment]
I have doubts about what to do with the Popup and also about refreshing all the interfaces that were added to the bridge, as currently just the interfaces configuration was written but the interface was not updated, and you could see them as not configured and things like that.

So i was in doubt if just duplicate current behavior or try to unify it with bonding.

File: src/include/network/lan/bridge.rb [hunk 23]
```diff
@@ -34,32 +34,97 @@ def initialize_network_lan_bridge(_include_target)
       textdomain "network"
     end
 
+    # It returns an empty string if given IP is 0.0.0.0, otherwise returns
+    # the given IP
+    #
+    # @param [String] ip
+    # @return [String] empty string if given ip is 0.0.0.0; given ip
+    # otherwise
+    def bridge_ip(ip)
```
Comment (mchf) [inline src/include/network/lan/bridge.rb]: it should be ```bridge_port_ip```

**Comment by teclator** [comment]
Have been replaced by #448
